src/Raven.Server/Documents/Replication/ReplicationTopologyDiscoveryHandler.cs

```diff
@@ -0,0 +1,127 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Raven.NewClient.Client.Http;
+using Raven.NewClient.Replication.Messages;
+using Raven.Server.Documents.TcpHandlers;
+using Raven.Server.Extensions;
+using Raven.Server.Json;
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.Documents.Replication
+{
+    public class ReplicationTopologyDiscoveryHandler
+    {
+        private readonly Guid _dbId;
+
+        public ReplicationTopologyDiscoveryHandler(Guid dbId)
+        {
+            _dbId = dbId;
+        }
+
+        public void AcceptIncomingConnectionAndRespond(TcpConnectionOptions tcp)
+        {
+            JsonOperationContext context;
+            using (tcp)
+            using (var multiDocumentParser = tcp.MultiDocumentParser)
+            using (tcp.DocumentDatabase.DocumentsStorage.ContextPool.AllocateOperationContext(out context))
+            {
+                TopologyDiscoveryHeader header;
+                using (var headerJson = multiDocumentParser.ParseToMemory("ReplicationTopologDiscovery/read-discovery-header"))
+                    header = JsonDeserializationServer.DiscoveryHeader(headerJson);
+
+                var outgoingHandlers = tcp.DocumentDatabase.DocumentReplicationLoader.Outgoing;
+
+                //if this node is a sink, no need to continue
+                if(outgoingHandlers.Count == 0)
+                {
+                    WriteDiscoveryResponse(tcp, context, new List<TopologyNode>(), TopologyDiscoveryResponse.ResponseType.Ok);
+                    return;
+                }
+
+                var token = tcp.DocumentDatabase.DatabaseShutdown;
+                using (var topologyDiscoverer = new ReplicationTopologyDiscoverer(
+                    outgoingHandlers,
+                    header.AlreadyVisitedDbIds,
+                    token))
+                {
+                    //if true, all adjacent nodes are already visited, so no need to continue
+                    //(filtering is done in a ctor of ReplicationTopologyDiscoverer, using header.AlreadyVisitedDbIds)
+                    if (topologyDiscoverer.DiscovererCount == 0)
+                    {
+                        WriteDiscoveryResponse(tcp, context, new List<TopologyNode>(), TopologyDiscoveryResponse.ResponseType.Ok);
+                        return;
+                    }
+
+                    var discoveryTask = topologyDiscoverer.DiscoverTopologyAsync();
+
+                    List<TopologyNode> nodes;
+                    TopologyDiscoveryResponse.ResponseType responseType;
+
+                    //if true, then the discovery is taking too long..
+                    //TODO : make topology discovery timeout configurable
+                    if (Task.WaitAny(discoveryTask, Task.Delay(TimeSpan.FromSeconds(60), token)) != discoveryTask.Id)
```

Comment:
This nee to be a parameter that is reduced throughout the network.
You are doing 60 seconds waits on each step, but this means that the first one can timeout because a totally different node is timing out.
You need to start with 60 seconds on the first node, then 50 seconds on the 2nd hop, etc.